openstack/02a-MAClow/IEEE802154_security.c

```diff
@@ -0,0 +1,860 @@
+                                                          msg->l2_frameType);
+
+   if (keyDescriptor==NULL){
+      openserial_printError(COMPONENT_SECURITY,ERR_SECURITY,
+                          (errorparameter_t)msg->l2_frameType,
+                          (errorparameter_t)0);
+      return E_FAIL;
+   }
+
+   for (j=0;j<16;j++){
+      key[j] = keyDescriptor->key[j];
+   }
+
+   uint8_t vectASN[5];
+   macFrameCounter_t l2_frameCounter;
+   ieee154e_getAsn(vectASN);//gets asn from mac layer.
+   if (frameCounterSuppression == 0){//the frame Counter is carried in the frame
+      //save the frame counter of the current frame
+      l2_frameCounter.bytes0and1 = vectASN[0]+256*vectASN[1];
+      l2_frameCounter.bytes2and3 = vectASN[2]+256*vectASN[3];
+      l2_frameCounter.byte4 = vectASN[4];
+
+      IEEE802154_security_getFrameCounter(l2_frameCounter,
+                                         msg->l2_FrameCounter);
+   } //otherwise the frame counter is not in the frame
+
+   //nonce creation
+   memset(&nonce[0], 0, 13);
+   //first 8 bytes of the nonce are always the source address of the frame
+   memcpy(&nonce[0],idmanager_getMyID(ADDR_64B)->addr_64b,8);
+
+   //Frame Counter (ASN)
+   for (i=0;i<5;i++){
+      nonce[8+i] = vectASN[i];
+   }
+
+   switch (msg->l2_securityLevel) {
+      case IEEE154_ASH_SLF_TYPE_MIC_32:  // authentication only cases
+      case IEEE154_ASH_SLF_TYPE_MIC_64:
+      case IEEE154_ASH_SLF_TYPE_MIC_128: 
+         a = msg->payload;             // first byte of the frame
+         len_a = msg->length;          // whole frame
+         m = &msg->payload[len_a];     // concatenate MIC at the end of the frame
+         len_m = 0;                    // length of the encrypted part
+         break;
+    case IEEE154_ASH_SLF_TYPE_CRYPTO_MIC32:  // authentication + encryption cases
+    case IEEE154_ASH_SLF_TYPE_CRYPTO_MIC64:
+    case IEEE154_ASH_SLF_TYPE_CRYPTO_MIC128:
+       a = msg->payload;             // first byte of the frame
+       m = msg->l2_payload;          // first byte where we should start encrypting (see 15.4 std)
+       len_a = m - a;                // part that is only authenticated is the difference of two pointers
+       len_m = msg->length - len_a;  // part that is encrypted+authenticated is the rest of the frame
+       break;
+    case IEEE154_ASH_SLF_TYPE_ONLYCRYPTO:    // encryption only
+       // unsecure, should not support it
+       return E_FAIL;
+    default:
+       // reject anything else
+       return E_FAIL;
+   }
+
+   // assert
+   if (len_a + len_m > 125) {
+      openserial_printError(COMPONENT_SECURITY,ERR_SECURITY,
+                           (errorparameter_t)msg->l2_frameType,
+                           (errorparameter_t)1);
+      return E_FAIL;
+   }
+
+   if (msg->l2_authenticationLength != 0) {
+      //update the length of the packet
+      packetfunctions_reserveFooterSize(msg,msg->l2_authenticationLength);
```

Comment:
I see your point. Could you check if commit "Shift payload entry point for length of the authentication tag." (d7fdce8bf903) addresses what you had in mind?

Note that latest commits will appear in this PR as soon as @ssciancalepore merges the PR. Until then you can see it here: ```https://github.com/malishav/openwsn-fw/commit/d7fdce8bf90385a66b1bf5189f22e5ed4cda5061```